contracts/AINFT721.sol

```diff
@@ -0,0 +1,304 @@
+    {
+        return super._isApprovedOrOwner(spender, tokenId);
+    }
+
+    function burn(uint256 tokenId) public override(ERC721Burnable) {
+        super.burn(tokenId);
+    }
+
+    ////
+    // URI & METADATA RELATED FUNCTIONS
+    ////
+
+    function _baseURI() internal view override(ERC721) returns (string memory) {
+        return baseURI;
+    }
+
+    function tokenURI(
+        uint256 tokenId
+    ) public view override(ERC721) returns (string memory) {
+        _requireMinted(tokenId);
+        uint256 currentVersion = _tokenURICurrentVersion[tokenId];
+
+        return tokenURIByVersion(tokenId, currentVersion);
+    }
+
+    /**
+     * @dev Returns the key for the metadata storage.
+     * @return The metadata storage key.
+     */
+    function _metadataStorageKey(
+        uint256 tokenId,
+        uint256 version
+    ) internal pure returns (bytes32) {
+        return keccak256(abi.encodePacked(tokenId.toString(), "AINFT delimeter", version));
+    }
+
+    /**
+     * @dev The metadata storage is a mapping of token ID to metadata.
+     * @return The metadata storage.
+     */
+    function metadataStorageByVersion(
+        uint256 tokenId,
+        uint256 version
+    ) public view returns (MetadataContainer memory) {
+    
+        bytes32 key = _metadataStorageKey(tokenId, version);
+        return _metadataStorage[key];
+    }
+
+    function tokenURICurrentVersion(
+        uint256 tokenId
+    ) public view returns (uint256) {
+        _requireMinted(tokenId);
+        return _tokenURICurrentVersion[tokenId];
+    }
+
+    function tokenURIByVersion(
+        uint256 tokenId,
+        uint256 uriVersion
+    ) public view returns (string memory) {
+        _requireMinted(tokenId);
+        if (uriVersion == 0) {
+            return string(abi.encodePacked(baseURI, tokenId.toString()));
+        } else {
+            MetadataContainer memory metadata = metadataStorageByVersion(tokenId, uriVersion);
+            return metadata.metadataURI;
+        }
+    }
+
+    function setBaseURI(
+        string memory newBaseURI
+    ) public onlyRole(DEFAULT_ADMIN_ROLE) returns (bool) {
+        require(
+            bytes(newBaseURI).length > 0,
+            "AINFT721::setBaseURI() - Empty newBaseURI"
+        );
+        require(
+            keccak256(bytes(newBaseURI)) != keccak256(bytes(baseURI)),
+            "AINFT721::setBaseURI() - Same newBaseURI as baseURI"
+        );
+
+        baseURI = newBaseURI;
+        //TODO: emit BatchMetadataUpdate(start, end). Consider after updateTokenURI() calls.
+        return true;
+    }
+
+    /**
+     * @dev version up & upload the metadata. You should call this function externally when the token is updated.
+     */
+    function updateTokenURI(
+        uint256 tokenId,
+        string memory newTokenURI
+    ) external returns (bool) {
+        require(
+            (isApprovedOrOwner(tx.origin, tokenId) ||
+             PAYMENT_PLUGIN == _msgSender()),
```

Comment:
If we already set PAYMENT_PLUGIN, then isApprovedOrOwner should be ignored. right? or I can call it free??

Reply:
`tx.origin `represents the source of sender. So when A call `AINPayment::executeUpdate()` and `executeUpdate() `calls `AINFT721::updateTokenURI()`, `tx.origin` is A and `msg.sender` is `AINPayment` contract.

Thus, `isApprovedOrOwner(tx.origin, tokenId)` can check the transaction sender is the owner of tokenId even if `PAYMENT_PLUGIN` is already set.

Reply:
The purpose of design is that `updateTokenURI` should ONLY be called by `AINPayment::executeUpdate()` to pay AIN.

Reply:
Then adding `require(PAYMENT_PLUGIN, "payment plugin required"` would be okay. right?

Reply:
Right. I'll add that part.

Reply:
Q) Should we update by calling AINFT721 after payment in AINPayment? (AINPayment -> AINFT721)
Can't we just call the AINPayment in AINFT721::updateTokenURI? (AINFT721 -> AINPayment)

If you make it flow like (AINFT721 -> AINPayment), 

* advantage: making it easy to set up payments.
1. deploy AINFT721
2. deploy AINPayment
3. set AINPayment Address at AINFT721

* disadvantage: only using one payment contract when updating.
However, I think it doesn't matter because there is only one payment contract set up now, too.

ps) I'm solidity baby, so this may be bullshit.